frontend/projects/upgrade/src/app/features/dashboard/home/components/metrics/metrics.component.ts

```diff
@@ -385,28 +321,170 @@ export class MonitoredMetricsComponent implements OnInit, OnChanges, OnDestroy {
     return metric;
   }
 
-  checkQueryNameError(queryName){
+  checkMetricKeyRequiredError(metricKeyMissing: Boolean){
+    if (metricKeyMissing) {
+      this.queryMetricKeyError.push(true);
+    }
+  }
+
+  checkQueryNameRequiredError(queryName: any){
     if (queryName == null || queryName === '') {
       this.queryNameError.push(true);
     }
   }
 
+  checkStatisticRequiredError(statistic: any) {
+    if (statistic == null || statistic === '') {
+      this.queryStatisticError.push(true);
+    }
+  }
+
+  checkComparisonStatisticRequiredError(compareFn: any, compareValue: any) {
+    if (compareFn == null || compareFn === '' || compareValue == null || compareValue === '') {
+      this.queryComparisonStatisticError.push(true);
+    }
+  }
+
+  ManageKeysControl(queryIndex: number, keyIndex: number) {
+    let keysArray;
+    if (queryIndex == null) {
+      queryIndex = this.queryIndex;
+    }
+    // Prepare filteredMetrics for each query and its keys for experimentInfo while in edit Mode
+    if (this.experimentInfo) {
+      if (this.experimentInfo.queries.length > 0) {
```

Comment:
This block looks like a duplicate on both sides of the `if/else`? I don't see the difference:

```typescript
keysArray = this.queries.at(queryIndex).get('keys') as FormArray;
        this.filteredMetrics$[keyIndex] = keysArray.at(keyIndex).get('metricKey').valueChanges
        .pipe(
          startWith<string>(''),
          map(key => {
            if (keyIndex - 1 >= 0) {
              const { metricKey } = keysArray.at(keyIndex - 1).value;
              this.options = metricKey.children;
            }
            return key ? this._filter(key) : this.options ? this.options.slice() : [];
          })
        );
```

Reply:
Again nice catch. Actually, earlier there was a difference in both sides of the `if/else` but with time changes were made upon debugging and finally it would have turned out to be exact same. My bad to have skipped this refactoring. I will do that.  Thanks :)